XPNet.CLR/Scenery/XPlaneScenery.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace XPNet
+{
+	/// <summary>
+	/// Provides access to the X-Plane scenery API, which allows working with the scenery.
+	/// On one hand, it provides probes for locating the physical scenery mesh (get the Y-coordinate of the ground for 
+	/// a defined X-Z coordinate).
+	/// Furthermore, it allows loading, drawing and managing scenery objects.
+	/// </summary>
+	public interface IXPlaneScenery
+	{
+		IXPProbe CreateProbe();
+
+		IXPSceneryObject LoadObject(string path);
+	}
+
+	internal class XPlaneScenery : IXPlaneScenery
+	{
+		public IXPProbe CreateProbe()
+		{
+			return new XPProbe(XPLMProbeType.xplm_ProbeY);
+		}
+
+		public IXPSceneryObject LoadObject(string path)
+		{
+			return new XPSceneryObject(path);
+		}
+	}
+
+	public interface IXPSceneryObject : IDisposable
+	{
+		void Draw(int lighting, int earthRelativ, XPLMDrawInfo_t[] drawInfos);
+	}
+
+	internal unsafe class XPSceneryObject : IXPSceneryObject
+	{
+		private void* m_objectRef;
+
+		public XPSceneryObject(string path)
+		{
+			m_objectRef = PluginBridge.ApiFunctions.XPLMLoadObject(path);
+			if (m_objectRef == null) throw new System.IO.FileNotFoundException();
+		}
+
+		public unsafe void Draw(int lighting, int earthRelative, XPLMDrawInfo_t[] drawInfos)
+		{
+			fixed (XPLMDrawInfo_t* drawInfosP = &drawInfos[0])
+				PluginBridge.ApiFunctions.XPLMDrawObjects(m_objectRef, drawInfos.Length, drawInfosP, lighting, earthRelative);
+		}
+
+		public void Dispose()
+		{
+			PluginBridge.ApiFunctions.XPLMUnloadObject(m_objectRef);
+		}
+	}
+
+
+	public interface IXPProbe : IDisposable
+	{
+		IXPProbeResult ProbeTerrainXYZ(float inX, float inY, float inZ);
+	}
+
+	public interface IXPProbeResult
+	{
+		float LocationX { get; }
+
+		float LocationY { get; }
+
+		float LocationZ { get; }
+
+		float NormalX { get; }
+
+		float NormalY { get; }
+
+		float NormalZ { get; }
+
+		float VelocityX { get; }
+
+		float VelocityY { get; }
+
+		float VelocityZ { get; }
+
+		int IsWet { get; }
+
+		XPLMProbeResult Result { get; }
+
+	}
+
+	internal class XPProbeResult : IXPProbeResult
+	{
+		private XPLMProbeInfo_t _probeInfo;
+		private XPLMProbeResult _result;
+
+		internal unsafe XPProbeResult(void* probeRef, float inX, float inY, float inZ)
+		{
+			_probeInfo = new XPLMProbeInfo_t();
+			_probeInfo.structSize = System.Runtime.InteropServices.Marshal.SizeOf(typeof(XPLMProbeInfo_t));
```

Comment:
Changed XPLMProbeInfo_t and XPLMDrawInfo_t to be classes instead of structs. The size of the structure is now initialized in the respective static constructor, and copied into the structure at object instantiation.

Reply:
I've never tried using StructLayout on a class before (as opposed to a struct).  I checked the docs and it looks like it's valid -- I bet it works as long as you are always creating the instance in your code, and never trying to map an instance created by native code.  So this is fine -- but if we ever needed to map to an instance created by native code so that we needed to turn this into a struct again, what I was intending to say is that we can make the Marshal.SizeOf call in a static constructor of _XPProbeResult_.